x-pack/plugins/security/public/management/roles/edit_role/privileges/es/index_privilege_form.tsx

```diff
@@ -50,9 +52,16 @@ interface State {
   grantedFields: string[];
   exceptedFields: string[];
   documentQuery?: string;
+  isFieldListLoading: boolean;
+  flsOptions: string[];
 }
 
 export class IndexPrivilegeForm extends Component<Props, State> {
+  // This is distinct from the field within `this.state`.
+  // We want to make sure that only one request for fields is in-flight at a time,
+  // and relying on state for this is error prone.
+  private isFieldListLoading: boolean = false;
```

Comment:
__question:__ just for my own education, would you mind explaining a bit when storing this flag only is `state` leads to errors?

Reply:
React can decide to batch/defer state updates, so calling `this.setState()` doesn't guarantee that your state will be updated within a specific period of time. Having this private field outside of state works around this by giving us a synchronously updating field that always reflects reality.